### DISCUSSION TITLE
Функция TryCatch и передача данных после отката

### DISCUSSION THREAD

**Discussion by erpic-ru** [totumonline/totum-ru-issues-and-discussions#38] [Как это сделать?] [answered]
Всем привет!

Задача следующая:

Есть простая таблица учета операций поступления и списания материалов с отдельной графой в виде накопительного регистра, в которой также рассчитывается остаток на складе после операции. Если после операции значение остатка становится меньше 0 нужно:
1. Показать ошибку, 
2. Откатить операцию 
3. Передать данные этой операции (материал, заказ, количество...) в другую простую таблицу.

У меня получается, но что-то одно. Либо ошибка с откатом, либо данные пишутся, а операция проходит

Подскажите, пожалуйста, как именно можно использовать для этого функцию tryCatch или через что-то другое сделать?

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Вот это бомбический вопрос!

Например ставим полю с количество код-действия по добавлению:

```
=catch: if(condition: $#nfv < 0; then: $err)

err: errorException(text: "Zero error")

catch: set(table: $#ntn; field: 'h_catch' = $@zero)
```

**Нашли, что `errorException` написан с ошибкой, поэтому до версии 2.* использовать `errorExeption`.**

Он проверяет неотрицательность и выводит ошибку. В `catch` выполняется нужное действие. Что бы записать вразумительную ошибку, можно юзать процессные переменные. Например:

```
a1=: procVar(name: "zero"; value: "test_value")

a2= : insert(table: $#ntn; field: 'test' = -1)

```

Это код кнопки, которая добавляет строку с непроходным значением.

Мы можем перед выполнением добавления строки записать все необходимые данные в процессную переменную. Доступ к ней будет у всех кодов, которые выполняются в цепочке. После того, как процесс PHP завершится данные этой переменной очищаются. Ну и между пользователями и разными действиями одного пользователя она не пересекается, тк изолирована конкретным процессом PHP.

И в `catch` мы используем данные из этой переменной.
